File: DSA/Queue/Code/PriorityQueue.py
# Priority Queue in Python

class Element:
    def __init__(self, data, priority):
        self.data = data
        self.priority = priority

class PriorityQueue:
    def __init__(self):
        self.arr = []
    
    def is_empty(self):
        return len(self.arr) == 0
    
    def is_full(self):
        return len(self.arr) == 100
    
    def enqueue(self, data, priority):
        if self.is_full():
            print("Priority Queue overflow")
            return
        
        elem = Element(data, priority)
        i = len(self.arr) - 1
        self.arr.append(elem)
        
        while i >= 0 and self.arr[i].priority < priority:
            self.arr[i + 1] = self.arr[i]
            i -= 1
        self.arr[i + 1] = elem
        
        print(f"Enqueued: data={data}, priority={priority} (size={len(self.arr)})")
    
    def dequeue(self):
        if self.is_empty():
            print("Priority Queue underflow")
            return -1
        data = self.arr[0].data
        print(f"Dequeued: data={data}, priority={self.arr[0].priority} (size={len(self.arr) - 1})")
        
        self.arr.pop(0)
        return data
    
    def display(self):
        if self.is_empty():
            print("Priority Queue is empty")
            return
        print("Priority Queue (data -> priority):")
        for elem in self.arr:
            print(f"  {elem.data} -> {elem.priority}")

# Test
pq = PriorityQueue()

print("=== Priority Queue ===\n")

pq.display()

print("\nEnqueuing elements:")
pq.enqueue(10, 2)
pq.enqueue(20, 5)
pq.enqueue(30, 1)
pq.enqueue(40, 3)

pq.display()

print("\nDequeuing (highest priority first):")
pq.dequeue()
pq.dequeue()

pq.display()

print("\nComplexity Analysis:")
print("Enqueue: O(n) - insertion in sorted order")
print("Dequeue: O(n) - removal with shifting")
print("Peek: O(1)")
print("Space: O(n)")
print("Note: Using linked list or heap improves Enqueue to O(log n)")

File: DSA/Queue/Code/PriorityQueue.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_SIZE 100

// Priority Queue in C (Max Heap)
typedef struct {
    int data;
    int priority;
} Element;

typedef struct {
    Element arr[MAX_SIZE];
    int size;
} PriorityQueue;

PriorityQueue* createPriorityQueue() {
    PriorityQueue* pq = (PriorityQueue*)malloc(sizeof(PriorityQueue));
    pq->size = 0;
    return pq;
}

bool isEmpty(PriorityQueue* pq) {
    return pq->size == 0;
}

bool isFull(PriorityQueue* pq) {
    return pq->size == MAX_SIZE;
}

void enqueue(PriorityQueue* pq, int data, int priority) {
    if (isFull(pq)) {
        printf("Priority Queue overflow\n");
        return;
    }
    
    int i = pq->size - 1;
    while (i >= 0 && pq->arr[i].priority < priority) {
        pq->arr[i + 1] = pq->arr[i];
        i--;
    }
    
    pq->arr[i + 1].data = data;
    pq->arr[i + 1].priority = priority;
    pq->size++;
    printf("Enqueued: data=%d, priority=%d (size=%d)\n", data, priority, pq->size);
}

int dequeue(PriorityQueue* pq) {
    if (isEmpty(pq)) {
        printf("Priority Queue underflow\n");
        return -1;
    }
    int data = pq->arr[0].data;
    printf("Dequeued: data=%d, priority=%d (size=%d)\n", data, pq->arr[0].priority, pq->size - 1);
    
    for (int i = 0; i < pq->size - 1; i++) {
        pq->arr[i] = pq->arr[i + 1];
    }
    pq->size--;
    return data;
}

void display(PriorityQueue* pq) {
    if (isEmpty(pq)) {
        printf("Priority Queue is empty\n");
        return;
    }
    printf("Priority Queue (data -> priority):\n");
    for (int i = 0; i < pq->size; i++) {
        printf("  %d -> %d\n", pq->arr[i].data, pq->arr[i].priority);
    }
}

int main() {
    PriorityQueue* pq = createPriorityQueue();
    
    printf("=== Priority Queue ===\n\n");
    
    display(pq);
    
    printf("\nEnqueuing elements:\n");
    enqueue(pq, 10, 2);
    enqueue(pq, 20, 5);
    enqueue(pq, 30, 1);
    enqueue(pq, 40, 3);
    
    display(pq);
    
    printf("\nDequeuing (highest priority first):\n");
    dequeue(pq);
    dequeue(pq);
    
    display(pq);
    
    printf("\nComplexity Analysis:\n");
    printf("Enqueue: O(n) - insertion in sorted order\n");
    printf("Dequeue: O(n) - removal with shifting\n");
    printf("Peek: O(1)\n");
    printf("Space: O(n)\n");
    printf("Note: Using linked list or heap improves Enqueue to O(log n)\n");
    
    free(pq);
    return 0;
}
